src/Contract/Test/Assert.purs

```diff
@@ -0,0 +1,600 @@
+        pure success
+
+  go :: ContractAssertion a
+  go = foldr wrapAssertion contract assertions
+
+tellFailure
+  :: ContractAssertionFailure -> ContractAssertion Unit
+tellFailure failure = do
+  ask >>= liftEffect <<< Ref.modify_ (Cons failure)
+
+checkNewUtxosAtAddress
+  :: forall (a :: Type)
+   . Labeled Address
+  -> TransactionHash
+  -> (Array TransactionOutputWithRefScript -> ContractAssertion a)
+  -> ContractAssertion a
+checkNewUtxosAtAddress addr txHash check =
+  lift (utxosAt $ unlabel addr) >>= \utxos ->
+    check $ Array.fromFoldable $ Map.values $
+      Map.filterKeys (\oref -> (unwrap oref).transactionId == txHash) utxos
+
+-- | Sets a limit on `ExUnits` budget. All ExUnits values of all submitted transactions are combined. Transactions that are constructed, but not submitted, are not considered.
+-- | The execution of the `Contract` will not be interrupted in case the `ExUnits` limit is reached.
+checkExUnitsNotExceed
+  :: forall (a :: Type)
+   . ExUnits
+  -> ContractCheck a
+checkExUnitsNotExceed maxExUnits contract = do
+  (ref :: Ref ExUnits) <- liftEffect $ Ref.new { mem: zero, steps: zero }
+  let
+    submitHook :: Transaction -> Effect Unit
+    submitHook tx = do
+      let
+        (newExUnits :: ExUnits) = sum $ tx ^..
+          _witnessSet
+            <<< _redeemers
+            <<< non []
+            <<< traversed
+            <<< to (unwrap >>> _.exUnits)
+      Ref.modify_ (add newExUnits) ref
+
+    setSubmitHook :: ContractEnv -> ContractEnv
+    setSubmitHook =
+      prop (Proxy :: Proxy "hooks") <<< prop (Proxy :: Proxy "onSubmit")
+        -- Extend a hook action if it exists, or set it to `Just submitHook`
+        %~ maybe (Just submitHook)
+          \oldHook -> Just \tx -> do
+            -- ignore possible exception from the old hook
+            void $ try $ oldHook tx
+            submitHook tx
+
+    finalize :: ContractAssertion Unit
+    finalize = do
+      exUnits <- liftEffect $ Ref.read ref
+      assertContract (MaxExUnitsExceeded maxExUnits exUnits)
+        (maxExUnits >= exUnits)
+
+  pure (mapReaderT (local setSubmitHook) contract /\ finalize)
+
+valueAtAddress'
+  :: Labeled Address
+  -> ContractAssertion Value
+valueAtAddress' = map (foldMap (view (_output <<< _amount))) <<< lift
+  <<< utxosAt
+  <<< unlabel
+
+-- | Arguments are:
+-- |
+-- | - a labeled address
+-- | - a callback that implements the assertion, accepting `Contract` execution
+-- |   result, and values (before and after)
+assertValueDeltaAtAddress
+  :: forall (a :: Type)
+   . Labeled Address
+  -> (a -> Value -> Value -> ContractAssertion Unit)
+  -> ContractCheck a
+assertValueDeltaAtAddress addr check contract = do
+  valueBefore <- valueAtAddress' addr
+  ref <- liftEffect $ Ref.new Nothing
+  let
+    finalize = do
+      valueAfter <- valueAtAddress' addr
+      liftEffect (Ref.read ref) >>= case _ of
+        Nothing -> pure unit -- tellFailure $ CustomFailure "Contract did not run"
```

Comment:
This comment is slightly misleading, since AFAIU the result of the contract isn't stored only in case of an exception, so the contract did actually run, but without success. Although this presupposition may be false with user-written assertions.